### ISSUE TITLE
Best way to handle a queue not known at the start of initial playback with callbacks between source loading

### ISSUE THREAD

**Issue by drhill** [google/ExoPlayer#4612]
I have multiple questions that are all mostly related for cleaning up my AndroidTV app.  I'm using Exoplayer 2 (a few weeks old) with the LeanbackPlayerAdapter and a PlaybackSupportFragment.  Mostly testing on Android O with multiple Nvidia Shield setups (and some Nexus Player but the performance on that device in the last 1.5+ years [prior to O and not fixed in it] is terrible as it needs to be rebooted daily).  The app plays videos (mostly MKVs) from an internal server via HTTP over the LAN.  

I load up a source URI in an ExtractorSource and start playback.  While playback is going I create a queue in the background which videos which would either take place before the currently playing video or after it.  Could be 0 videos before/after, could be 100s of videos.  Then I have hacked up some logic which will skip chapters if available (known from the metadata received from the server I am using to store the video files as ExoPlayer doesn't seem to care about chapters, MKV or otherwise) when you hit the previous/next buttons on the remote or OSD if they are available and if they aren't (or the current position is near the duration boundaries I've set for myself) the app moves on by calling setDataSource for the next video.  It's hacky and I wanted to start using MediaSession/MediaSessionConnector as I try to cleanup this cludge.

The questions I have are:

1.  What's the best way to handle a situation where I may need to grow a queue of 1, possible in both directions?  I'm assuming something with ConcatenatingMediaSource and MediaSession so Google Assistant/PIP title's can be supported.  Questions #3 and #4 below are going to have some relevance to this regarding callbacks when switching from one video to the next.

2.  Any suggestions for handling chapters?  I get a headache trying to read and figure out the source for MediaPeriods, Windows, etc in ExoPlayer. 

3.  To get MKV files with DTS HD tracks to passthrough properly (with DTSHD/MA support) I have a hack that doesn't feel right, but works.  I created a class that inherits from MediaCodecAudioRenderer and overload onInputFormatChanged(Format format).  Since I externally have extra metadata of the video I'm playing, if the format passed in matches a track in my metadata which is a DTSHD track I rewrite the format variable to use a mime type of MimeTypes.AUDIO_DTS_HD... it feels dirty to use this.  Is there a better way?

4.  Refresh rate switching.  Most of the videos I play use a refresh rate of 23.976hz.  On my test setups without external receiver's attached things seem to work fine however I load videos and start playback.  On my main testing area I have a Denon receiver and a Sony SXRD projector.  On this setup I'll get an error unless I wait to load the datasource and begin playback until after the refresh rate is done plus a few seconds. That is I have wait for a DisplayManager.DisplayListener.onDisplayChanged callback plus around 3-4 seconds.  I've adapted the code to do that but it's odd that I have no issues with changing the refresh rate in my fragment's onResume after coming back from the screensaver after pausing my video and releasing the screenlock.  I return back, change the refresh rate, then hit play to resume the video and acquire the screen lock with no issues.  But changing the display rate while initially setting the datasource will throw an error about the audiotrack failing to write.  I'm assuming it has to do with the HDMI source chicanery going on at the time but I'm quite certain at the time I have PlayWhenReady set to false.  Is there any advice that can be given for smoother display rate changes?

**Comment by erdemguven** [comment]
@ojw28 could you look at this?

**Comment by ojw28** [comment]
1. Yes. Use `ConcatenatingMediaSource`. I'm not sure there's anything more to it than this?
1. I would expect chapter information to be handled outside of ExoPlayer (e.g. by application code seeking to the start of the chapter, assuming that's a common use case). You'd need to provide more information about your use case, and what problems you're encountering, for us to provide more specific advice.
1. What mime type does the `Format` have before you re-write it, and why? If it's a bug in our MKV extractor then we should fix it to output a `Format` that has the correct mime type. If it's a bug in the content then the content should be fixed. Please clarify. Thanks!
1. I agree with your diagnosis (i.e. the issue is probably caused by the change in HDMI mode invalidating the `AudioTrack` in the underlying platform). Having `setPlayWhenReady` set to false is not sufficient to prevent audio being written to the `AudioTrack`; ExoPlayer will still prime the `AudioTrack` with data so that it's ready to play immediately when `setPlayWhenReady` is toggled to `true`. If possible, you should change the display mode before calling `prepare` on the player (I'm not sure how practical this will be for you).

**Comment by ojw28** [comment]
For (3), it looks like this is discussed in https://github.com/google/ExoPlayer/issues/2147 as well. Basically someone needs to improve the MKV extractor so that it's able to detect DTSHD and output a `Format` with the correct mime type.

**Comment by drhill** [comment]
Over the weekend I changed all my logic to use mediasession and mediasession connector.  In the short term I decided against using ConcatenatingMediaSource because I wasn't sure how well switching refresh rates would be with that implementation.  I just made my own QueueNavigator that accessed my applications queue instead of Exoplayer's TimelineQueueNavigator. 

For chapters just wasn't sure if this is something that ExoPlayer could/should handle currently.  There is no issue with the handling chapters that way.  

I already handle the refresh rate switching in the way you suggest, though I still have to add a delay after onDisplayChanged returns for with one of my setups.  I'll double check to see why I have to for that one setup.  Perhaps the multi stage chain of receiver and projector some extra callbacks I'm not accounting for.

I'll look around to see if there is a way to properly detect DTSHD in MKVs.

Thanks.
